### PR TITLE
Implement Rewards Account for Payouts and Update ChainSpec

### PR DESCRIPTION
**Pull request by asiniscalchi**
## **Type**
enhancement

___

## **Description**
- Introduced `TransferFromRewardsAccount` for handling rewards payout from a specified account.
- Updated `ChainSpec` to include rewards account and inflation configuration.
- Refactored Parachain Staking to use the new rewards account for payouts.
- Removed the `BlockRewardsHandler` pallet from the runtime.
- Updated network specifications for Laos Sun v2 Testnet with adjusted inflation.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>laos.rs</strong><dd><code>Update ChainSpec with Rewards Account and Inflation Config</code></dd></summary>
<hr>

node/src/chain_spec/laos.rs
<li>Updated <code>ChainSpec</code> to include <code>rewards_account</code> in <code>testnet_genesis</code>.<br> <li> Removed duplicate <code>get_collator_keys_from_seed("Alice")</code> entries.<br> <li> Adjusted <code>testnet_genesis</code> to use a single <code>stakers</code> vector including <br><code>AccountId</code>, <code>AuraId</code>, and <code>Balance</code>.<br> <li> Introduced <code>rewards_account</code> and <code>inflation_config</code> in <code>parachain_staking</code> <br>configuration.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-2ccef9e2947edd74c86ff7b5a814c38a09079830181b3f5d86bb098384e7a868">+24/-8</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>parachain_staking.rs</strong><dd><code>Configure Parachain Staking to Use Rewards Account</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

runtime/laos/src/configs/parachain_staking.rs
- Updated `PayoutReward` to use `TransferFromRewardsAccount`.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-260a47c783c16f4ca58a07f9d05998e138f42333640d8c0e48b7df110a0143d8">+3/-4</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Remove Block Rewards Handler Pallet from Runtime</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

runtime/laos/src/lib.rs
- Removed `BlockRewardsHandler` pallet.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-003c87018a12c01c5266d5e1f3aba274f64cc6651fdb6b34abb88a6d1a10d073">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>laossun_v2.plain.json</strong><dd><code>Update Network Specs for Laos Sun v2 Testnet</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

specs/laossun_v2.plain.json
<li>Updated network name and ID to <code>Laos Sun v2 Testnet</code>.<br> <li> Adjusted annual inflation to 7.5%.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-debca9d073d93caee55c917099f3fe1e917e046bb9540697b02ea7a2d720bb01">+7/-9</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Refactor Parachain Staking to Use Rewards Account for Payouts</code></dd></summary>
<hr>

pallets/parachain-staking/src/lib.rs
<li>Added <code>rewards</code> module import.<br> <li> Removed <code>mint_collator_reward</code> function.<br> <li> Changed <code>PayoutReward</code> trait implementation to use <br><code>TransferFromRewardsAccount</code>.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86e">+25/-17</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transfer_from_rewards_account.rs</strong><dd><code>Implement Rewards Transfer from Specified Account</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs
<li>Introduced <code>TransferFromRewardsAccount</code> struct implementing <br><code>PayoutReward</code>.<br> <li> Implemented logic to transfer rewards from a specified rewards <br>account.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/449/files#diff-996f4e8d6d54b24b8894159eb2a791e90668a4f7c842a3787bf919d60230a8c4">+212/-0</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		) {
+			Ok(_) => Ok(amount),
+			Err(DispatchError::Arithmetic(ArithmeticError::Underflow)) => Ok(Zero::zero()),
+			Err(e) => Err(e),
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn send_collator_reward(
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:58]:
```suggestion
	pub fn send_collator_rewards(
```

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		_round_idx: RoundIndex,
+		collator_id: T::AccountId,
+		amt: BalanceOf<T>,
+	) -> Weight {
+		// Check if the collator's account exists; return early if not.
+		if !frame_system::Account::<T>::contains_key(&collator_id) {
+			return Weight::zero(); // TODO
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:65]: what is the TODO?

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+			Err(e) => Err(e),
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn send_collator_reward(
+		_round_idx: RoundIndex,
+		collator_id: T::AccountId,
+		amt: BalanceOf<T>,
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:61]:
```suggestion
		amount: BalanceOf<T>,
```
to keep the naming consistent

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+		// Proceed with the transfer and handle the result.
+		let transfer_result = T::Currency::transfer(
+			&rewards_account,
+			&collator_id,
+			amt,
+			ExistenceRequirement::KeepAlive,
+		);
+
+		match transfer_result {
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:84]: we are not interested in unwrapping value of `Ok()` here so, maybe it's better to:

```rs
if let Err(e) = transfer_result {
...
} else {
...
}
```

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+	#[test]
+	fn test_payout_with_no_rewards_account_should_do_nothing() {
+		ExtBuilder::default().build().execute_with(|| {
+			let delegator = 0;
+			let amount = 100;
+
+			let _ = pallet_balances::Pallet::<Test>::deposit_creating(&delegator, 1);
+
+			assert_ok!(
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:152]: probably better to use `assert_noop` here

File: pallets/parachain-staking/src/lib.rs [hunk 6]
```diff
@@ -689,15 +698,29 @@ pub mod pallet {
...
 		fn build(&self) {
+			RewardsAccount::<T>::set(self.rewards_account.clone());
+
 			assert!(self.blocks_per_round > 0, "Blocks per round must be > 0");
-			<InflationConfig<T>>::put(self.inflation_config.clone());
+
+			// Set inflation configuration
+			let mut inflation_config = self.inflation_config.clone();
+			// if all the round values are 0, derive them from the annual values
+			if inflation_config.round.min.is_zero() &&
```
Comment (dastanfv) [inline pallets/parachain-staking/src/lib.rs:716]: can you explain again why this check is needed? do we allow round values that are non-zero as an input?

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs (hunk 2 above)
Comment (asiniscalchi) [reply pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:65]: handled here: https://github.com/freeverseio/laos/pull/466

Comment (dastanfv) [reply pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:65]: oh, I meant that there is nothing to do here, we return zero weight if rewards account does not exist, no?

File: pallets/parachain-staking/src/rewards/minting_rewards.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
+use crate::{traits::PayoutReward, BalanceOf, Config, Event, Pallet, RoundIndex, WeightInfo};
+use frame_support::{
+	pallet_prelude::Weight,
+	traits::{tokens::currency::Currency, Imbalance},
+};
+use sp_runtime::DispatchError;
+
+pub struct MintingRewards;
```
Comment (dastanfv) [inline pallets/parachain-staking/src/rewards/minting_rewards.rs:8]:
```suggestion
pub struct MintRewards;
```
also would suggest renaming this to be consistent with `TransferFromRewardsAccount`:

File: pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs (hunk 5 above)
Comment (asiniscalchi) [reply pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:150]: `asset_noop` is used in error cases

Comment (dastanfv) [reply pallets/parachain-staking/src/rewards/transfer_from_rewards_account.rs:150]: ah yeah, then `assert_storage_noop`

File: pallets/parachain-staking/src/lib.rs (hunk 6 above)
Comment (asiniscalchi) [reply pallets/parachain-staking/src/lib.rs:716]: it's the best way I found to communicate the user that those parameters are not under his control that avoid a refactoring

Comment (dastanfv) [reply pallets/parachain-staking/src/lib.rs:716]: but now it is possible to pass any non-zero value for round, so _user_ is in control? if you just always derive round values from annual you don't need to make any refactoring of the struct, no?

**Pull request merged**